### PR TITLE
feat: [IDP-1043]: Add scorecard validation & optimise entity facets API

### PR DESCRIPTION
**Pull request by vigneswara-propelo**
## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/52085)

### CODE REVIEW

**Review by vikyathharekal — COMMENTED** [review]
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/52085)* status: 1 file(s) to review. 0 of 2 approvals obtained. 

___
*[`idp-service/src/main/java/io/harness/idp/scorecard/scorecardchecks/service/ScorecardServiceImpl.java` line 63 at r1](https://reviewable.io/reviews/harness/harness-core/52085#-NchYDc9Ch0n7ICTcMbB:-NchYDc9Ch0n7ICTcMbC:blbvsgy) ([raw file](https://github.com/harness/harness-core/blob/c000dab55044d229b4d92804f79d36f576430c8f/idp-service/src/main/java/io/harness/idp/scorecard/scorecardchecks/service/ScorecardServiceImpl.java#L63)):*
> ```Java
>   private static final String CATALOG_API =
>       "%s/idp/api/catalog/entity-facets?filter=kind=%s&facet=%s&facet=%s&facet=%s&facet=%s";
>   private static final String TYPE_FILTER = "spec.type";
> ```

Can we have these constants hardcoded in the CATALOG_API itself if there is nothing dynamic in it?
`"%s/idp/api/catalog/entity-facets?filter=kind=%s&facet=spec.type&facet=relations.ownedBy&facet=metadata.tags&facet=spec.lifecycle"`

**Review by vigneswara-propelo — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/52085)* status: 1 file(s) to review. 1. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`idp-service/src/main/java/io/harness/idp/scorecard/scorecardchecks/service/ScorecardServiceImpl.java` line 63 at r1](https://reviewable.io/reviews/harness/harness-core/52085#-NchYDc9Ch0n7ICTcMbB:-NchaGQv34m-DkOIib17:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/c000dab55044d229b4d92804f79d36f576430c8f/idp-service/src/main/java/io/harness/idp/scorecard/scorecardchecks/service/ScorecardServiceImpl.java#L63)):*
<details><summary><i>Previously, vikyathharekal (Vikyath Harekal) wrote…</i></summary><blockquote>

Can we have these constants hardcoded in the CATALOG_API itself if there is nothing dynamic in it?
`"%s/idp/api/catalog/entity-facets?filter=kind=%s&facet=spec.type&facet=relations.ownedBy&facet=metadata.tags&facet=spec.lifecycle"`
</blockquote></details>

Done.

**Comment by vigneswara-propelo** [comment]
trigger uts

File: idp-service/src/main/java/io/harness/idp/scorecard/scorecardchecks/service/ScorecardServiceImpl.java [hunk 1]
```diff
@@ -158,6 +162,14 @@ public ScorecardDetailsResponse getScorecardDetails(String accountIdentifier, St
     return ScorecardDetailsMapper.toDTO(scorecardEntity, checkEntityMap, accountIdentifier);
   }
 
+  private void validateScorecard(ScorecardDetailsRequest scorecardDetailsRequest) {
```
Comment (sathish-soundarapandian) [inline idp-service/src/main/java/io/harness/idp/scorecard/scorecardchecks/service/ScorecardServiceImpl.java:165]: nit: can be renamed as validateScorecardSaveRequest
nit: multiple if conditions can be merged to one in the validate function

**Review by vikyathharekal — DISMISSED** [review]
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/52085)* status: 1 file(s) to review. 1 of 2 approvals obtained.

File: idp-service/src/main/java/io/harness/idp/scorecard/scorecardchecks/service/ScorecardServiceImpl.java (hunk 1 above)
Comment (vigneswara-propelo) [reply idp-service/src/main/java/io/harness/idp/scorecard/scorecardchecks/service/ScorecardServiceImpl.java]: done

**Review by vikyathharekal — APPROVED** [review]
> Reviewed 1 of 1 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/52085)* status: 1 file(s) to review. 1 of 2 approvals obtained.

**Comment by vigneswara-propelo** [comment]
trigger uts

**Review by sathish-soundarapandian — APPROVED** [review]

**Comment by sathish-soundarapandian** [comment]
trigger review

**Pull request merged**